### PR TITLE
Expand api

### PR DESCRIPTION
**Pull request by Gusted**
Expand our API + document them + Providing an OAuth option to access our API on behalf of users.

### CODE REVIEW

**Comment by vednoc** [comment]
This will definitely take quite a while to test.

**Comment by Gusted** [comment]
> This will definitely take quite a while to test.

use the callback_helper that's provided in the seeded OAuth will help some time by receiving the access_token.

File: database/database.go [hunk 1]
```diff
@@ -182,6 +183,18 @@ func seed() {
...
+			UserID:       1,
+			Name:         "USw integration",
+			Description:  "Just some integration",
+			Scopes:       []string{"style", "user"},
+			ClientID:     "publicccc_client",
+			ClientSecret: "secreettUwU",
+			RedirectURI:  "https://gusted.xyz/callback_helper/usw",
+		},
+	}
+
```
Comment (vednoc) [inline database/database.go:197]: Make this only available in dev environment?

File: handlers/api/user.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+		"data": fiber.Map{
+			"username":  user.Username,
+			"email":     user.Email,
+			"biography": user.Biography,
+			"role":      roleToString(user.Role),
+			"socials": fiber.Map{
+				"github":   user.Socials.Github,
+				"gitlab":   user.Socials.Gitlab,
+				"codeberg": user.Socials.Codeberg,
+			},
```
Comment (vednoc) [inline handlers/api/user.go:53]: Shouldn't this use a User model?

File: handlers/api/user.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	return c.JSON(fiber.Map{
+		"data": fiber.Map{
+			"username":  user.Username,
+			"biography": user.Biography,
+			"role":      roleToString(user.Role),
+			"socials": fiber.Map{
+				"github":   user.Socials.Github,
+				"gitlab":   user.Socials.Gitlab,
+				"codeberg": user.Socials.Codeberg,
+			},
```
Comment (vednoc) [inline handlers/api/user.go:84]: Shouldn't this use a User model?

File: handlers/server.go [hunk 4]
```diff
@@ -97,15 +101,20 @@ func Initialize() {
 	if config.Production {
 		app.Use(limiter.New(limiter.Config{Max: 300}))
 	}
-	app.Use(jwt.New())
+	app.Use(jwtware.New("user", func(t *jwt.Token) (interface{}, error) {
+		if t.Method.Alg() != jwtware.SigningMethod {
+			return nil, fmt.Errorf("unexpected jwt signing method=%v", t.Header["alg"])
+		}
+		return jwtware.JWTSigningKey, nil
+	}))
```
Comment (vednoc) [inline handlers/server.go:109]: Can this be extracted from here?

File: handlers/server.go [hunk 5]
```diff
@@ -97,15 +101,20 @@ func Initialize() {
...
 
 	app.Get("/", core.Home)
 	app.Get("/search", core.Search)
 	app.Get("/login", user.LoginGet)
 	app.Post("/login", user.LoginPost)
 	app.Get("/register", user.RegisterGet)
 	app.Post("/register", user.RegisterPost)
-	app.Get("/oauth/:type", user.AuthLoginGet)
+	app.Get("/oauth_login/:type", user.AuthLoginGet)
```
Comment (vednoc) [inline handlers/server.go:117]: Remove `_login`? This also goes for `btn_auth` template.

File: handlers/server.go [hunk 6]
```diff
@@ -118,27 +127,42 @@ func Initialize() {
...
 	v1.Get("/search/:query", api.GetSearchResult)
 	v1.Get("/callback/:rcode", api.CallbackGet)
+	v1.Get("/user", api.ProtectedAPI, api.UserGet)
+	v1.Get("/user/:identifier", api.SpecificUserGet)
+	v1.Get("/styles", api.ProtectedAPI, api.StylesGet)
+	v1.Post("/style/new", api.ProtectedAPI, api.NewStyle)
+	v1.Post("/style/:id", api.ProtectedAPI, api.StylePost)
+	v1.Delete("/style/:id", api.ProtectedAPI, api.DeleteStyle)
+
+	oauthV1 := app.Group("/oauth")
```
Comment (vednoc) [inline handlers/server.go:160]: Prefix this with `/api`?

File: database/database.go (hunk 1 above)
Comment (Gusted) [reply database/database.go:197]: Isn't it already? It will only be seeded once you set DB_DROP env

Comment (vednoc) [reply database/database.go:197]: Oh yeah, I forgot. 😅

**Review by vednoc — APPROVED** [review]

**Pull request merged**